### PR TITLE
New swimming effects

### PR DESCRIPTION
**Pull request by gromchek**
`Swim like Tommy` - player takes full health damage in ~5 sec
`Disable swimming` - disable controls in water

### CODE REVIEW

File: src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include "util/EffectBase.h"
+
+class DisableSwimmingEffect : public EffectBase
+{
+public:
+    void
+    OnProcessScripts (EffectInstance *inst) override
+    {
+        CPlayerPed *player = FindPlayerPed ();
+        if (player && !FindPlayerVehicle (-1, false))
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp]: Please use an early return here - `if (!player || FindPlayerVehicle(-1, false)) return;

File: src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+public:
+    void
+    OnProcessScripts (EffectInstance *inst) override
+    {
+        CPlayerPed *player = FindPlayerPed ();
+        if (player && !FindPlayerVehicle (-1, false))
+        {
+            CPad *pad = player->GetPadFromPlayer ();
+            if (!pad)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp]: Please have the `return;` be on the same line without the curly brackets

File: src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            {
+                return;
+            }
+
+            bool cond = player->m_pIntelligence->GetTaskSwim ();
+            cond |= player->m_nPhysicalFlags.bSubmergedInWater;
+
+            if (cond)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/DisableSwimmingEffect.cpp]: You might be able to just do `pad->DisablePlayerControls = cond;` here

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+class SwimLikeTommyEffect : public EffectBase
+{
+    int wait = 0;
+
+public:
+    void
+    OnProcessScripts (EffectInstance *inst) override
+    {
+        CPlayerPed *player = FindPlayerPed ();
+        if (player && !FindPlayerVehicle (-1, false))
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: Same here, early return please

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            CPad *pad = player->GetPadFromPlayer ();
+            if (!pad)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: Same as in the other file, return on the same line

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                return;
+            }
+
+            bool cond = player->m_pIntelligence->GetTaskSwim ();
+            cond |= player->m_nPhysicalFlags.bSubmergedInWater;
+
+            if (cond)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: And same as in the other file, too - Can just do `pad->DisablePlayerControls = cond;` probably

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                pad->DisablePlayerControls = false;
+            }
+        }
+    }
+
+    void
+    OnTick (EffectInstance *inst) override
+    {
+        wait -= (int) GenericUtil::CalculateTick ();
+        if (wait > 0)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: Return on the same line, please

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            return;
+        }
+
+        CPlayerPed *player = FindPlayerPed ();
+        if (!player)
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: Return on the same line, please

File: src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        if (cond)
+        {
+            player->m_fHealth -= player->m_fMaxHealth * 0.01f;
+        }
+
+        wait = 50;
+    }
+};
+
+DEFINE_EFFECT (SwimLikeTommyEffect, "effect_swim_like_tommy", GROUP_HEALTH);
```
Comment (Lordmau5) [inline src/gtasa/effects/custom/player/SwimLikeTommyEffect.cpp]: Seeing as you modify the health *and* controls, I'd prefer if you could do `GROUP_HEALTH | GROUP_CONTROLS` so it's in both groups

**Review by Lordmau5 — CHANGES_REQUESTED** [review]

**Comment by Lordmau5** [comment]
Oh also I forgot to add that in the review, you can credit yourself at the top of the files if you want
Honestly, I'd even prefer that :)

Something like `// Effect by gromcheck` or similar

**Comment by gromchek** [comment]
Fixed

**Comment by Lordmau5** [comment]
Dunno why it failed with the compilation... I managed to compile it just fine locally

I'll get this merged and also push the effects in the GUI, but I'm not sure when I'll be able to get out a new version with these

**Comment by Lordmau5** [comment]
Alright so I went ahead and checked if the effects even work properly just now LMAO
Should've done that before...

Either way - `Swim Like Tommy` drains the player's health when they are in the water, but neither of the 2 effects actually disable swimming controls, or controls at all, for me... Which is interesting. I know other effects like `Let's Take A Break` do it properly so I'm pretty confused myself ...

A-ha! I found the culprit!
```cpp
if (!player || !FindPlayerVehicle (-1, false)) return;
```
So you're checking if the player is false / null (which makes sense) **OR** if the Vehicle is false / null.
This means the effect will only work when the player is in a vehicle AND submerged in water
I'll get that fixed in a commit in a bit :)

I'd also like to remove the "submerged in water" check and only go for "GetTaskSwim" instead. Would that be okay with you?
Because otherwise the moment CJ get's his feet wet (as in, just standing in a puddle) he won't be able to move (or in `Swim Like Tommy`'s case, will lose health)

**Pull request merged**